### ISSUE TITLE
Openssl compilation for ARM(Linux arria10 4.14.130-ltsi-altera armv7l GNU/Linux)

### ISSUE THREAD

**Issue by ankit7gup** [openssl/openssl#14438]
Hi,

We are compiling Openssl to be used with Libssh. However the Libssh compilation is failing as it is unable to get a reference for some functions defined under Openssl. 

We want to cross compile OpenSSL for Target: Linux arria10 4.14.130-ltsi-altera #1 SMP Thu Jan 9 10:55:28 UTC 2020 armv7l GNU/Linux

And we are using below mentioned commands to compile OpenSSL.

````
 ./Configure linux-elf no-asm shared --prefix=/usr/local --openssldir=/usr/local/ssl
 make 
 make install
 ````

 ##########Errors in Libssh compilation###############

````
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:17:39: error: expected declaration specifiers or ‘...’ before string constant
 static void *OPENSSL_zalloc(size_t num)
                                       ^
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:17:39: error: expected declaration specifiers or ‘...’ before numeric constant
 static void *OPENSSL_zalloc(size_t num)
                                       ^
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c: In function ‘RSA_set0_key’:
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:32:11: error: dereferencing pointer to incomplete type ‘RSA {aka struct rsa_st}’
     if ((r->n == NULL && n == NULL)
           ^~
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c: In function ‘RSA_get0_key’:
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:103:15: error: dereferencing pointer to incomplete type ‘RSA {aka const struct rsa_st}’
         *n = r->n;
               ^~
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c: In function ‘DSA_get0_pqg’:
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:134:15: error: dereferencing pointer to incomplete type ‘DSA {aka const struct dsa_st}’
         *p = d->p;
               ^~
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c: In function ‘DSA_set0_pqg’:
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:146:11: error: dereferencing pointer to incomplete type ‘DSA {aka struct dsa_st}’
     if ((d->p == NULL && p == NULL)
           ^~
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c: In function ‘DSA_SIG_get0’:
/root/ankit/libssh-0.9.5/src/libcrypto-compat.c:200:18: error: dereferencing pointer to incomplete type ‘DSA_SIG {aka const struct DSA_SIG_st}’
         *pr = sig->r;
````
##########################################
 Kindly help us with the compilation procedure or correct if we are missing something.
 
 Thanks!
 
 Regards,
Ankit

**Comment by mattcaswell** [comment]
This looks like a question that would be better targeted at the libssh team rather than us.

It looks to me that what is failing is libssh's own compatibility layer around OpenSSL's libcrypto. From the look of things it seems to be trying to provide a compatibility wrapper around OpenSSL 1.0.2 to provide replacement functions that aren't available in that version - but are only available in later versions. For example, the function `RSA_set0_key` is a function that is available in libcrypto in OpenSSL 1.1.0+, but is not available in OpenSSL 1.0.2.

The error `dereferencing pointer to incomplete type ‘RSA {aka struct rsa_st}` is typical of an application expecting to use the OpenSSL 1.0.2 API but is being compiled against OpenSSL 1.1.0+ headers.

So my guess is you have configured the libssh compilation to use OpenSSL 1.0.2, but provided headers for OpenSSL 1.1.0+. But, as I said, this is really a question for libssh.

**Comment by ankit7gup** [comment]
Hi Matt, 

First of all, Thanks for you quick response and analysis. 
With help of your observation, found out that LIBSSH even latest version(0.9.5) is compatible with OPEN SSL 1.0.2K version only and not the other ones.
Mentioning here, in case somebody might have similar query in future. This thread may help.

Thanks!
Ankit
